### PR TITLE
Add Kimi K3 SiTU dense MLP support

### PR DESCRIPTION
**Pull request by YangXu1990uiuc**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I added GitHub labels: one `cat-*`, one or more `mod-*`, and one `orig-*` (see [label list](https://github.com/NVIDIA/cudnn-frontend/labels)).
- [ ] I set the **Milestone** and **Projects** fields in the sidebar (required to merge; maintainers can set these for external contributions).

The milestone is set to **Frontend 1.29.0**. Projects remains unset because the current GitHub token does not have project scope.

## Affected area

Python API or bindings

### Downstream model impact (combined attribution)

On B200, an official-count Kimi-K3 BF16 fprop+bprop feature-substack integration carrying a patch-identical rebase of this SiTU implementation measured **1.3567x / 1.3238x / 1.2189x** at 8K/16K/32K, removing **26.3% / 24.5% / 18.0%** of the declared substack latency. The all-on arm also replaces KDA and all three short convolutions, while MLA remains unchanged, so this is not a marginal #807 or full-model result. The SiTU implementation is unchanged from the measured branch; the current head adds documentation, contract-focused test cleanup, and an upstream rebase. The table is historical combined attribution, not a fresh current-head model rerun.

- Add a distinct `cudnn.gemm.situ_mlp(x, Wg, Wu, Wd, *, beta=4.0, linear_beta=25.0)` entry point for the bounded SiTU MLP used by Kimi K3, while leaving the existing `swiglu_mlp(x, Wg, Wu, Wd)` API and behavior unchanged.
- Implement `beta * tanh(gate / beta) * sigmoid(gate)` and `linear_beta * tanh(up / linear_beta)` with BF16 linear boundaries and FP32 pointwise computation.
- Add exact SiTU backward, activation/beta-aware plan caches and autograd snapshots, partial-gradient handling, and a fail-closed guard that prevents the SwiGLU-only FROST backward from serving SiTU.
- Share private graph/autograd machinery between the two semantic entry points and keep lazy exports stable in either import order.

## Why

Kimi K3 uses bounded gate and up activations rather than ordinary SwiGLU. Treating the two formulas as one public operation with an activation switch would make it easier to select the wrong derivative or reuse a semantically incompatible plan. A separate semantic entry point preserves the existing SwiGLU contract and makes both SiTU beta values explicit plan inputs.

The formula follows the [official Moonshot Kimi K3 implementation](https://huggingface.co/moonshotai/Kimi-K3/blob/main/modeling_kimi_linear.py).

For inference, cuDNN fuses both BF16 input projections with the FP32 SiTU pointwise sequence and multiply into one runtime-fusion kernel, rounds the fused activation to BF16, and executes the down projection as a second kernel. The exact PyTorch composition launches the projections and pointwise sequence separately. The measured advantage therefore comes from fewer launches and less intermediate traffic under the same math and BF16 boundaries; graph construction and autotuning happen before steady-state replay.

## Related issues

None.

## API and compatibility impact

This adds one public Python API:

```python
cudnn.gemm.situ_mlp(x, Wg, Wu, Wd, *, beta=4.0, linear_beta=25.0)
```

- `x` is BF16 `[..., H]`; `Wg` and `Wu` are BF16 `[I, H]`; `Wd` is BF16 `[H, I]`. All tensors must be on the same CUDA device.
- `beta` and `linear_beta` are keyword-only positive finite Python numbers. The defaults match Kimi K3.
- Forward and backward are differentiable with respect to all four tensor inputs. SiTU backward uses the exact cuDNN derivative graph and deliberately does not use the current SwiGLU-only FROST backward.
- The fused forward currently requires SM100 and BF16. Unsupported architectures or dtypes fail instead of silently selecting different semantics.
- Each distinct beta pair owns separately autotuned plan/workspace cache entries for the process lifetime.
- Existing `cudnn.gemm.swiglu_mlp` callers are backward compatible.

## Testing

- Current review head `aafea264e`: changed-file pre-commit, Python compilation, and diff checks passed.
- NVIDIA B200 with cuDNN 9.26 at the production-identical prior head `070d3edd2`: the complete L0 file reported **31 passed, 2 skipped**; the skipped cases require more than one visible GPU.
- NVIDIA B200 at current head: the four public SiTU API/forward/backward/beta-contract tests reported **4 passed, 28 deselected**.
- Forward and all four gradients match an independent PyTorch reference. Coverage also includes alternate-beta semantics, partial gradients, checkpointing, CUDA graph capture/replay, stream and device behavior, and public input validation. The current cleanup removes assertions that monkeypatch or directly call the private FROST route.

Same-process interleaved CUDA-graph replay device intervals, inference only, 9 median samples with 40 iterations per sample:

| Kimi K3 role | Shape `(M, H, I)` | cuDNN replay | PyTorch replay | PyTorch / cuDNN | Relative L2 |
|---|---:|---:|---:|---:|---:|
| dense | `(512, 7168, 33792)` | `0.552705 ms` | `0.887115 ms` | `1.605x` | `7.44e-4` |
| shared expert | `(2048, 7168, 6144)` | `0.370443 ms` | `0.627354 ms` | `1.694x` | `6.64e-4` |
| routed-expert proxy | `(128, 3584, 3072)` | `0.020578 ms` | `0.043139 ms` | `2.096x` | `6.06e-4` |

These timings cover the complete three-GEMM SiTU inference operator, not a single kernel. They exclude first-call graph construction and autotuning. The routed row count is an operation proxy rather than assembled MoE dispatch, and the results do not claim full-model, training-step, or non-B200 speedup.

## Summary by CodeRabbit

- **New Features**
  - Added the `cudnn.gemm.situ_mlp` API for Kimi K3 bounded SiTU MLPs.
  - Supports configurable activation bounds with validated parameters.
  - Provides accurate forward and backward computations alongside existing SwiGLU functionality.
  - Automatically uses supported acceleration paths while preserving correct SiTU gradients.

- **Documentation**
  - Added integration guidance, performance measurements, and accuracy results for supported data types and expert configurations.

- **Tests**
  - Added coverage for SiTU forward/backward parity, parameter validation, API behavior, and acceleration-path handling.

### CODE REVIEW

File: python/cudnn/gemm/ops/swiglu_mlp.py [hunk 1]
```diff
@@ -73,6 +90,48 @@
...
+        positive_finite("beta", situ_beta),
+        positive_finite("linear_beta", situ_linear_beta),
+    )
+
+
+def _graph_scalar(g, value, name):
+    # Runtime-param scalars work with the op's cuDNN 9.x floor; compile-time
+    # constants need cuDNN >= 9.22. The graph owns the value, and the cache key
+    # owns its identity, so execute() has no extra tensor/scalar argument.
+    return g.tensor_scalar(float(value), cudnn.scalar_type.RUNTIME_PARAM, name=name)
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply python/cudnn/gemm/ops/swiglu_mlp.py:133]: Verified against the current public graph path: `cudnn.pygraph` is the Python-native `cudnn._pygraph.pygraph` class (wired in `python/cudnn/__init__.py`), and its `tensor_scalar(self, value, scalar_type, name="")` API explicitly accepts `name`. The two-argument C++ overload inspected here is used only during lowering, not by `_graph_scalar`. The existing call is therefore valid and is intentionally retained.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost,python_tests

**Review by yanqinz2 — APPROVED** [review]

**Comment by YangXu1990uiuc** [comment]
@coderabbitai review

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run python_tests,frost

**Comment by YangXu1990uiuc** [comment]
@coderabbitai review

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost,python_tests
